File: retaRust/python_arch_reference/reta_architecture/execution_network.py
from __future__ import annotations

"""Deterministic execution-network primitives for Reta runtime work.

This module deliberately sits *above* the mathematical layers.  Topologies,
presheaves, sheaves, morphisms, functors and natural transformations stay pure;
this layer only schedules and glues executable tasks that already have a clear
semantic owner.
"""

import heapq
import importlib
import multiprocessing
import os
from collections import deque
from dataclasses import dataclass, field
from queue import Queue
from threading import BoundedSemaphore
from typing import Any, Callable, Iterable, Mapping, MutableSequence, Sequence


def _available_worker_count() -> int:
    process_cpu_count = getattr(os, "process_cpu_count", None)
    if callable(process_cpu_count):
        try:
            value = process_cpu_count()
            if value:
                return max(1, int(value))
        except Exception:
            pass
    sched_getaffinity = getattr(os, "sched_getaffinity", None)
    if callable(sched_getaffinity):
        try:
            return max(1, len(sched_getaffinity(0)))
        except Exception:
            pass
    return max(1, os.cpu_count() or 1)


EXECUTION_NETWORK_PROCESS_CORES = _available_worker_count()


@dataclass(frozen=True)
class ExecutionNetworkConfig:
    """Configuration for deterministic task execution.

    ``use_processes`` selects multiprocessing rather than threads.  Worker counts
    are capped by both the configured maximum and the number of available tasks.
    """

    max_workers: int = EXECUTION_NETWORK_PROCESS_CORES
    queue_discipline: str = "fifo"
    use_processes: bool = False
    start_method: str | None = None
    preserve_input_order: bool = True
    bounded_queue_size: int | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "max_workers", max(1, int(self.max_workers or 1)))
        discipline = str(self.queue_discipline or "fifo").lower()
        if discipline not in {"fifo", "lifo", "priority"}:
            discipline = "fifo"
        object.__setattr__(self, "queue_discipline", discipline)
        if self.start_method in {"", "default", "none"}:
            object.__setattr__(self, "start_method", None)
        if self.bounded_queue_size is not None:
            object.__setattr__(self, "bounded_queue_size", max(1, int(self.bounded_queue_size)))

    def workers_for(self, task_count: int) -> int:
        return min(self.max_workers, max(1, int(task_count or 1)))

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "max_workers": self.max_workers,
            "resolved_available_workers": EXECUTION_NETWORK_PROCESS_CORES,
            "queue_discipline": self.queue_discipline,
            "use_processes": self.use_processes,
            "start_method": self.start_method,
            "preserve_input_order": self.preserve_input_order,
            "bounded_queue_size": self.bounded_queue_size,
        }


@dataclass(frozen=True)
class ExecutionTask:
    """One executable local task.

    ``callable_path`` is preferred for process execution because it is importable
    under both ``fork`` and ``spawn``.  The callable receives ``payload`` and
    returns a serialisable value.
    """

    index: int
    payload: Any
    operation: str = "identity"
    priority: int = 0
    callable_path: str | None = None
    metadata: Mapping[str, Any] = field(default_factory=dict)

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "index": self.index,
            "operation": self.operation,
            "priority": self.priority,
            "callable_path": self.callable_path,
            "metadata": dict(self.metadata),
        }


@dataclass(frozen=True)
class ExecutionResult:
    task_index: int
    value: Any
    operation: str
    metadata: Mapping[str, Any] = field(default_factory=dict)

    def snapshot(self) -> dict:
        try:
            value_len = len(self.value)
        except Exception:
            value_len = None
        return {
            "class": type(self).__name__,
            "task_index": self.task_index,
            "operation": self.operation,
            "value_type": type(self.value).__name__,
            "value_len": value_len,
            "metadata": dict(self.metadata),
        }


@dataclass(frozen=True)
class ExecutionRunResult:
    values: list[Any]
    results: tuple[ExecutionResult, ...]
    config: ExecutionNetworkConfig
    workers: int
    task_count: int
    queue_discipline: str
    mode: str

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "values": len(self.values),
            "results": [item.snapshot() for item in self.results],
            "config": self.config.snapshot(),
            "workers": self.workers,
            "task_count": self.task_count,
            "queue_discipline": self.queue_discipline,
            "mode": self.mode,
            "universal_property": "parallel_or_serial_task_cover_glues_to_the_same_ordered_result",
        }


class FifoTaskQueue:
    def __init__(self, tasks: Iterable[ExecutionTask] = ()) -> None:
        self._items = deque(tasks)

    def push(self, task: ExecutionTask) -> None:
        self._items.append(task)

    def pop(self) -> ExecutionTask:
        return self._items.popleft()

    def __len__(self) -> int:
        return len(self._items)


class LifoTaskStack:
    def __init__(self, tasks: Iterable[ExecutionTask] = ()) -> None:
        self._items = list(tasks)

    def push(self, task: ExecutionTask) -> None:
        self._items.append(task)

    def pop(self) -> ExecutionTask:
        return self._items.pop()

    def __len__(self) -> int:
        return len(self._items)


class PriorityTaskQueue:
    def __init__(self, tasks: Iterable[ExecutionTask] = ()) -> None:
        self._heap: list[tuple[int, int, ExecutionTask]] = []
        for task in tasks:
            self.push(task)

    def push(self, task: ExecutionTask) -> None:
        heapq.heappush(self._heap, (task.priority, task.index, task))

    def pop(self) -> ExecutionTask:
        return heapq.heappop(self._heap)[2]

    def __len__(self) -> int:
        return len(self._heap)


class ResourceSemaphore:
    """Small process-safe semaphore wrapper used only at resource boundaries."""

    def __init__(self, value: int) -> None:
        self.value = max(1, int(value or 1))
        self._semaphore = BoundedSemaphore(self.value)

    def acquire(self, timeout: float | None = None) -> bool:
        if timeout is None:
            return bool(self._semaphore.acquire())
        return bool(self._semaphore.acquire(timeout=timeout))

    def release(self) -> None:
        self._semaphore.release()

    def snapshot(self) -> dict:
        return {"class": type(self).__name__, "value": self.value}


class HalfDuplexChannel:
    """Request/response channel: one direction is active at a time."""

    def __init__(self, maxsize: int = 0) -> None:
        self.requests: Queue = Queue(maxsize=max(0, int(maxsize or 0)))
        self.responses: Queue = Queue(maxsize=max(0, int(maxsize or 0)))

    def send_request(self, message: Any) -> None:
        self.requests.put(message)

    def receive_request(self, timeout: float | None = None) -> Any:
        return self.requests.get(timeout=timeout)

    def send_response(self, message: Any) -> None:
        self.responses.put(message)

    def receive_response(self, timeout: float | None = None) -> Any:
        return self.responses.get(timeout=timeout)

    def snapshot(self) -> dict:
        return {"class": type(self).__name__, "requests": self.requests.qsize(), "responses": self.responses.qsize()}


class FullDuplexChannel:
    """Two independent queues for bidirectional supervisor/worker traffic."""

    def __init__(self, maxsize: int = 0) -> None:
        self.a_to_b: Queue = Queue(maxsize=max(0, int(maxsize or 0)))
        self.b_to_a: Queue = Queue(maxsize=max(0, int(maxsize or 0)))

    def send_a_to_b(self, message: Any) -> None:
        self.a_to_b.put(message)

    def receive_a_to_b(self, timeout: float | None = None) -> Any:
        return self.a_to_b.get(timeout=timeout)

    def send_b_to_a(self, message: Any) -> None:
        self.b_to_a.put(message)

    def receive_b_to_a(self, timeout: float | None = None) -> Any:
        return self.b_to_a.get(timeout=timeout)

    def snapshot(self) -> dict:
        return {"class": type(self).__name__, "a_to_b": self.a_to_b.qsize(), "b_to_a": self.b_to_a.qsize()}


@dataclass(frozen=True)
class ExecutionNetworkBundle:
    config: ExecutionNetworkConfig
    cpu_semaphore: ResourceSemaphore
    file_io_semaphore: ResourceSemaphore
    output_semaphore: ResourceSemaphore

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "category": "ExecutionNetworkCategory",
            "scheduler_category": "SchedulerCategory",
            "channel_category": "ChannelCategory",
            "config": self.config.snapshot(),
            "processor_cores": EXECUTION_NETWORK_PROCESS_CORES,
            "queues": ["FifoTaskQueue", "LifoTaskStack", "PriorityTaskQueue"],
            "channels": ["HalfDuplexChannel", "FullDuplexChannel"],
            "semaphores": {
                "cpu": self.cpu_semaphore.snapshot(),
                "file_io": self.file_io_semaphore.snapshot(),
                "output": self.output_semaphore.snapshot(),
            },
            "morphisms": [
                "enqueue_task",
                "dequeue_task",
                "dispatch_task",
                "collect_result",
                "deterministic_reduce",
                "acquire_resource",
                "release_resource",
                "send_message",
                "receive_message",
            ],
            "universal_property": "parallel_chunks_glue_deterministically_to_serial_result",
        }


def bootstrap_execution_network(config: ExecutionNetworkConfig | None = None) -> ExecutionNetworkBundle:
    config = config or ExecutionNetworkConfig()
    return ExecutionNetworkBundle(
        config=config,
        cpu_semaphore=ResourceSemaphore(config.max_workers),
        file_io_semaphore=ResourceSemaphore(max(1, min(4, config.max_workers))),
        output_semaphore=ResourceSemaphore(1),
    )


def _queue_for(config: ExecutionNetworkConfig, tasks: Sequence[ExecutionTask]):
    if config.queue_discipline == "lifo":
        return LifoTaskStack(tasks)
    if config.queue_discipline == "priority":
        return PriorityTaskQueue(tasks)
    return FifoTaskQueue(tasks)


def order_tasks(tasks: Sequence[ExecutionTask], config: ExecutionNetworkConfig) -> list[ExecutionTask]:
    queue = _queue_for(config, tasks)
    ordered: list[ExecutionTask] = []
    while len(queue) > 0:
        ordered.append(queue.pop())
    return ordered


def _resolve_callable(callable_path: str | None, handler: Callable[[Any], Any] | None = None) -> Callable[[Any], Any]:
    if handler is not None:
        return handler
    if callable_path:
        module_name, func_name = callable_path.split(":", 1)
        module = importlib.import_module(module_name)
        return getattr(module, func_name)
    return _builtin_operation


def _builtin_operation(payload: Any) -> Any:
    return payload


def _task_worker(task: ExecutionTask) -> ExecutionResult:
    handler = _resolve_callable(task.callable_path)
    return ExecutionResult(
        task_index=task.index,
        value=handler(task.payload),
        operation=task.operation,
        metadata=task.metadata,
    )


def _run_serial(tasks: Sequence[ExecutionTask], handler: Callable[[Any], Any] | None = None) -> list[ExecutionResult]:
    results: list[ExecutionResult] = []
    for task in tasks:
        resolved_handler = _resolve_callable(task.callable_path, handler)
        results.append(
            ExecutionResult(
                task_index=task.index,
                value=resolved_handler(task.payload),
                operation=task.operation,
                metadata=task.metadata,
            )
        )
    return results


def deterministic_reduce(results: Sequence[ExecutionResult], *, preserve_input_order: bool = True) -> list[Any]:
    ordered = sorted(results, key=lambda item: item.task_index) if preserve_input_order else list(results)
    return [item.value for item in ordered]


def execute_tasks_deterministically(
    tasks: Sequence[ExecutionTask],
    *,
    handler: Callable[[Any], Any] | None = None,
    config: ExecutionNetworkConfig | None = None,
) -> ExecutionRunResult:
    """Execute tasks via FIFO/LIFO/priority scheduling and deterministic reduce.

    Process mode requires importable ``task.callable_path`` entries.  Passing a
    local ``handler`` intentionally forces serial mode to avoid hidden thread-like
    behaviour or non-portable process pickling.
    """

    config = config or ExecutionNetworkConfig()
    tasks = tuple(tasks)
    if not tasks:
        return ExecutionRunResult([], (), config, workers=0, task_count=0, queue_discipline=config.queue_discipline, mode="empty")
    scheduled = order_tasks(tasks, config)
    use_processes = bool(config.use_processes and handler is None and any(task.callable_path for task in scheduled))
    workers = config.workers_for(len(scheduled)) if use_processes else 1
    if use_processes:
        start_method = config.start_method
        ctx = multiprocessing.get_context(start_method) if start_method else multiprocessing.get_context()
        with ctx.Pool(processes=workers) as pool:
            results = pool.map(_task_worker, scheduled)
        mode = "processes"
    else:
        results = _run_serial(scheduled, handler=handler)
        mode = "serial"
    values = deterministic_reduce(results, preserve_input_order=config.preserve_input_order)
    return ExecutionRunResult(
        values=values,
        results=tuple(results),
        config=config,
        workers=workers,
        task_count=len(scheduled),
        queue_discipline=config.queue_discipline,
        mode=mode,
    )

File: retaRust/crates/reta_architecture/src/execution_network.rs
//! Named Rust execution-network adapter for `execution_network.py`.
//!
//! Stage 8 put the concrete queue/channel code in `dataflow.rs`.  This module
//! restores the Python architecture name `execution_network` and adds a bridge
//! plan that can be exported without running work.

use serde::{Deserialize, Serialize};

pub use crate::dataflow::{
    bootstrap_execution_network, deterministic_reduce, execute_tasks_deterministically,
    ExecutionNetworkBundle, ExecutionNetworkConfig, ExecutionResult, ExecutionRunResult,
    ExecutionTask, FifoTaskQueue, FullDuplexChannel, HalfDuplexChannel,
    LifoTaskStack, PriorityTaskQueue, ResourceSemaphore,
};
use crate::dataflow::{order_tasks, DataflowDiscipline};


/// Python-architecture compatibility: available worker count used by
/// `execution_network.py`.  This mirrors Python's guarded CPU discovery with
/// Rust's portable available_parallelism fallback.
pub fn _available_worker_count() -> usize {
    std::thread::available_parallelism()
        .map(|value| value.get())
        .unwrap_or(1)
        .max(1)
}

/// Normalize config after construction, matching Python `__post_init__`.
pub fn __post_init__(mut config: ExecutionNetworkConfig) -> ExecutionNetworkConfig {
    config.max_workers = config.max_workers.max(1);
    config
}

/// Default constructor mirror for Python dataclass `__init__`.
pub fn __init__() -> ExecutionNetworkConfig {
    ExecutionNetworkConfig::default()
}

/// Pick the configured queue discipline and return the scheduled task order.
pub fn _queue_for<T: Clone>(
    config: &ExecutionNetworkConfig,
    tasks: &[ExecutionTask<T>],
) -> Vec<ExecutionTask<T>> {
    order_tasks(tasks, config)
}

/// Rust-local identity operation used when no importable Python callable exists.
pub fn _builtin_operation<T: Clone>(payload: &T) -> T {
    payload.clone()
}

/// In Rust the callable is already typed; resolving it is the identity over the
/// handler reference.  The name is kept to preserve the Python architecture map.
pub fn _resolve_callable<T, U, F>(handler: &F) -> &F
where
    F: Fn(&T) -> U,
{
    handler
}

/// Execute one task with the identity operation.
pub fn _task_worker<T: Clone>(task: &ExecutionTask<T>) -> ExecutionResult<T> {
    ExecutionResult {
        task_index: task.index,
        value: task.payload.clone(),
        operation: task.operation.clone(),
        metadata: task.metadata.clone(),
    }
}

/// Serial worker path used by the deterministic execution-network morphism.
pub fn _run_serial<T, U, F>(tasks: &[ExecutionTask<T>], handler: &F) -> Vec<ExecutionResult<U>>
where
    T: Clone,
    U: Clone,
    F: Fn(&T) -> U,
{
    order_tasks(tasks, &ExecutionNetworkConfig::default())
        .iter()
        .map(|task| ExecutionResult {
            task_index: task.index,
            value: handler(&task.payload),
            operation: task.operation.clone(),
            metadata: task.metadata.clone(),
        })
        .collect()
}

pub fn push<T>(queue: &mut FifoTaskQueue<T>, task: ExecutionTask<T>) {
    queue.push(task);
}

pub fn pop<T>(queue: &mut FifoTaskQueue<T>) -> Option<ExecutionTask<T>> {
    queue.pop()
}

pub fn acquire(semaphore: &ResourceSemaphore) {
    semaphore.acquire();
}

pub fn release(semaphore: &ResourceSemaphore) {
    semaphore.release();
}

pub fn send_request<T>(channel: &HalfDuplexChannel<T>, message: T) {
    channel.send_request(message);
}

pub fn receive_request<T>(channel: &HalfDuplexChannel<T>) -> T {
    channel.receive_request()
}

pub fn send_response<T>(channel: &HalfDuplexChannel<T>, message: T) {
    channel.send_response(message);
}

pub fn receive_response<T>(channel: &HalfDuplexChannel<T>) -> T {
    channel.receive_response()
}

pub fn send_a_to_b<T>(channel: &FullDuplexChannel<T>, message: T) {
    channel.send_a_to_b(message);
}

pub fn receive_a_to_b<T>(channel: &FullDuplexChannel<T>) -> T {
    channel.receive_a_to_b()
}

pub fn send_b_to_a<T>(channel: &FullDuplexChannel<T>, message: T) {
    channel.send_b_to_a(message);
}

pub fn receive_b_to_a<T>(channel: &FullDuplexChannel<T>) -> T {
    channel.receive_b_to_a()
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ExecutionNetworkPlan {
    pub task_count: usize,
    pub scheduled_indices: Vec<usize>,
    pub queue_discipline: String,
    pub workers: usize,
    pub preserve_input_order: bool,
    pub bounded_queue_size: Option<usize>,
    pub mode: String,
    pub universal_property: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ExecutionNetworkGate {
    pub gate_id: String,
    pub allowed_disciplines: Vec<String>,
    pub requires_deterministic_reduce: bool,
    pub status: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ExecutionNetworkBridgeSnapshot {
    pub class: String,
    pub queue_disciplines: Vec<String>,
    pub gates: usize,
    pub default_workers: usize,
    pub universal_property: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ExecutionNetworkBridgeBundle {
    pub default_config: ExecutionNetworkConfig,
    pub gates: Vec<ExecutionNetworkGate>,
}

impl ExecutionNetworkBridgeBundle {
    pub fn plan_for_tasks<T: Clone>(&self, tasks: &[ExecutionTask<T>]) -> ExecutionNetworkPlan {
        let scheduled = order_tasks(tasks, &self.default_config);
        ExecutionNetworkPlan {
            task_count: tasks.len(),
            scheduled_indices: scheduled.iter().map(|task| task.index).collect(),
            queue_discipline: self.default_config.queue_discipline.as_str().to_string(),
            workers: self.default_config.workers_for(tasks.len()),
            preserve_input_order: self.default_config.preserve_input_order,
            bounded_queue_size: self.default_config.bounded_queue_size,
            mode: if self.default_config.max_workers > 1 && tasks.len() > 1 { "threaded-plan" } else { "serial-plan" }.to_string(),
            universal_property: "parallel_chunks_glue_deterministically_to_serial_result".to_string(),
        }
    }

    pub fn plan_for_indices(&self, indices: &[usize]) -> ExecutionNetworkPlan {
        let tasks = indices.iter().map(|idx| ExecutionTask::new(*idx, *idx)).collect::<Vec<_>>();
        self.plan_for_tasks(&tasks)
    }

    pub fn snapshot(&self) -> ExecutionNetworkBridgeSnapshot {
        ExecutionNetworkBridgeSnapshot {
            class: "ExecutionNetworkBridgeBundle".to_string(),
            queue_disciplines: vec!["fifo".to_string(), "lifo".to_string(), "priority".to_string()],
            gates: self.gates.len(),
            default_workers: self.default_config.max_workers,
            universal_property: "queue_stack_priority_schedulers_share_one_deterministic_reduce".to_string(),
        }
    }
}

pub fn bootstrap_execution_network_bridge(config: Option<ExecutionNetworkConfig>) -> ExecutionNetworkBridgeBundle {
    let default_config = config.unwrap_or_else(|| ExecutionNetworkConfig { max_workers: 1, queue_discipline: DataflowDiscipline::Fifo, use_processes: false, start_method: None, preserve_input_order: true, bounded_queue_size: None });
    ExecutionNetworkBridgeBundle {
        default_config,
        gates: vec![
            ExecutionNetworkGate { gate_id: "serial-vs-threaded-order".to_string(), allowed_disciplines: vec!["fifo".to_string(), "lifo".to_string(), "priority".to_string()], requires_deterministic_reduce: true, status: "required".to_string() },
            ExecutionNetworkGate { gate_id: "bounded-queue-semaphore".to_string(), allowed_disciplines: vec!["fifo".to_string()], requires_deterministic_reduce: true, status: "optional".to_string() },
            ExecutionNetworkGate { gate_id: "bidirectional-channel-roundtrip".to_string(), allowed_disciplines: vec!["fifo".to_string()], requires_deterministic_reduce: false, status: "required-for-prompt".to_string() },
        ],
    }
}

pub fn execution_network_plan_for_indices(indices: &[usize], discipline: DataflowDiscipline) -> ExecutionNetworkPlan {
    let config = ExecutionNetworkConfig { queue_discipline: discipline, ..ExecutionNetworkConfig::default() };
    bootstrap_execution_network_bridge(Some(config)).plan_for_indices(indices)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn priority_plan_orders_indices_but_preserves_glue_law() {
        let cfg = ExecutionNetworkConfig { max_workers: 2, queue_discipline: DataflowDiscipline::Priority, use_processes: false, start_method: None, preserve_input_order: true, bounded_queue_size: None };
        let tasks = vec![ExecutionTask::new(0, "a").with_priority(2), ExecutionTask::new(1, "b").with_priority(1)];
        let plan = bootstrap_execution_network_bridge(Some(cfg)).plan_for_tasks(&tasks);
        assert_eq!(plan.scheduled_indices, vec![1, 0]);
        assert!(plan.universal_property.contains("deterministically"));
    }
}
